File: javascript/lib/ComputerStatusExamples.js
/*
 * Copyright 2019 Trend Micro.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * Obtains agent and appliance status for all comptuters
 * and provides the results as comma-separated values.
 * @param {object} api The api module.
 * @param {String} apiVersion The api version to use.
 * @return {Promise} A promise that contains a string of values in CSV format.
 */
exports.getComputerStatuses = function(api, apiVersion) {
  return new Promise((resolve, reject) => {
    // Obtains a list of all computers
    const getListOfComputers = () => {
      const computersApi = new api.ComputersApi();
      return computersApi.listComputers(apiVersion, { overrides: false });
    };
    // Add column titles to comma-separated values string
    let csv = "Host Name, Agent or Appliance, Status, Status Messages, Tasks\r\n";
    // Get the computers and iterate them
    getListOfComputers()
      .then(computers => {
        for (const computer of computers.computers) {
          const computerInfo = []; // Stores computer status information
          // Report on computers with no agent or appliance
          if ((computer.agentFingerPrint === undefined) & (computer.applianceFingerPrint === undefined)) {
            // Hostname and protection type
            computerInfo.push(computer.hostName);
            computerInfo.push("None");
            // Agent/appliance status and status message
            computerInfo.push("No agent/appliance");
            computerInfo.push(
              computer.computerStatus.agentStatus !== undefined ? computer.computerStatus.agentStatusMessages : ""
            );
            // Add to CSV string
            csv += formatForCSV(computerInfo);
          } else {
            // Report on problem agents and appliances
            const agentIsActive = computer.computerStatus.agentStatus == api.ComputerStatus.AgentStatusEnum.active;
            const applianceIsActive =
              computer.computerStatus.applianceStatus == api.ComputerStatus.ApplianceStatusEnum.active;
            if (computer.agentFingerPrint !== undefined && !agentIsActive) {
              // Agent is installed but not active
              computerInfo.push(computer.hostName);
              computerInfo.push("Agent");
              computerInfo.push(
                computer.computerStatus.agentStatus !== undefined ? computer.computerStatus.agentStatus : ""
              );
              computerInfo.push(
                computer.computerStatus.agentStatusMessages !== undefined
                  ? computer.computerStatus.agentStatusMessages
                  : ""
              );
              computerInfo.push(computer.tasks !== undefined ? computer.tasks.agentTasks : "");
              // Add to CSV string
              csv += formatForCSV(computerInfo);
              computerInfo.lenght = 0;
            }
            if (computer.applianceFingerPrint !== undefined && !applianceIsActive) {
              // Appliance is installed but not active
              computerInfo.push(computer.hostName);
              computerInfo.push("Appliance");
              computerInfo.push(
                computer.computerStatus.applianceStatus !== undefined ? computer.computerStatus.applianceStatus : ""
              );
              computerInfo.push(
                computer.computerStatus.applianceStatusMessages !== undefined
                  ? computer.computerStatus.applianceStatusMessages
                  : ""
              );
              computerInfo.push(computer.tasks !== undefined ? computer.tasks.applianceTasks : "");
              // Add to CSV string
              csv += formatForCSV(computerInfo);
            }
          }
        }
        resolve(csv);
      })
      .catch(error => {
        reject(error);
      });
  });
};

exports.getAntiMalwareStatusForComputers = function(api, apiVersion) {
  return new Promise((resolve, reject) => {
    // Obtains a list of all computers
    const getListOfComputers = () => {
      const computersApi = new api.ComputersApi();
      return computersApi.listComputers(apiVersion, { overrides: false });
    };
    // Add column titles to comma-separated values string
    let csv = "Host Name, Module State, Agent or Appliance, Status, Status Message\r\n";
    // Get the computers and iterate them
    getListOfComputers()
      .then(computers => {
        for (const computer of computers.computers) {
          let moduleInfo = [];
          const agentStatus = computer.antiMalware.moduleStatus.agentStatus;
          const applianceStatus = computer.antiMalware.moduleStatus.applianceStatus;

          // Agents that are not active for the module
          if (agentStatus !== undefined && agentStatus !== api.ComputerModuleStatus.AgentStatusEnum.active) {
            // Hostname
            moduleInfo.push(computer.hostName);
            // Module state
            moduleInfo.push(computer.antiMalware.state);
            // Agent status and status message
            moduleInfo.push("Agent");
            moduleInfo.push(agentStatus);
            moduleInfo.push(computer.antiMalware.moduleStatus.agentStatusMessages);

            // Add to the CSV string
            csv += formatForCSV(moduleInfo);
            moduleInfo.length = 0;
          }
          // Appliances that are not active for the module
          if (applianceStatus !== undefined && agentStatus !== api.ComputerModuleStatus.ApplianceStatusEnum.active) {
            // Hostname
            moduleInfo.push(computer.hostName);
            // Module state
            moduleInfo.push(computer.antiMalware.state);
            // Agent status and status message
            moduleInfo.push("Appliance");
            moduleInfo.push(applianceStatus);
            moduleInfo.push(computer.antiMalware.moduleStatus.applianceStatusMessages);

            // Add to the CSV string
            csv += formatForCSV(moduleInfo);
          }
        }
        resolve(csv);
      })
      .catch(error => {
        reject(error);
      });
  });
};

// PRIVATE METHODS //

/*
 * Converts an array of string values into a string of comma-separated values.
 * @param {Array} values The array of values.
 * @return {String} The string of comma-separated values.
 */
function formatForCSV(values) {
  let csvLine = "";
  for (let i = 0; i < values.length; i++) {
    csvLine += '"' + values[i] + '"';
    if (i != values.length - 1) {
      csvLine += ",";
    } else {
      csvLine += "\r\n";
    }
  }
  return csvLine;
}

/*
 * Obtains certain Anti-Malware properties for a computer.
 * @param {object} api The api module.
 * @param {Number} computerID The ID of the computer.
 * @param {String} apiVersion The api version to use.
 * @return {Promise} A promise that contains on object of Anti-Malware properties.
 */

exports.checkAntiMalware = function(api, computerID, apiVersion) {
  return new Promise((resolve, reject) => {
    let amStatus, computer;

    // Gets the computer object from Deep Security Manager
    const getComputer = () => {
      const computersApi = new api.ComputersApi();
      return computersApi.describeComputer(computerID, apiVersion, {
        overrides: false
      });
    };

    // Gets the malware scan configuration for a computer
    const getMalwareScanConfig = computerObj => {
      const realTimeScanConfigID = computerObj.antiMalware.realTimeScanConfigurationID;
      const amConfigsApi = new api.AntiMalwareConfigurationsApi();
      return amConfigsApi.describeAntiMalware(realTimeScanConfigID, apiVersion);
    };

    getComputer()
      .then(computerObj => {
        computer = computerObj;
        amStatus = getAntiMalwareInfo(computerObj);
        if (computerObj.antiMalware.realTimeScanConfigurationID !== 0) {
          getMalwareScanConfig(computerObj)
            .then(scanConfig => {
              amStatus.directoriesToScan = scanConfig.directoriesToScan;
              resolve(amStatus);
            })
            .catch(error => {
              reject(error);
            });
        } else {
          resolve(amStatus);
        }
      })
      .catch(error => {
        reject(error);
      });
  });
};

// PRIVATE METHODS //

// Retrieves certain Anti-Malware properties from a computer object
function getAntiMalwareInfo(computer) {
  const status = {};
  status.name = computer.hostName;
  status.state = computer.antiMalware.state;
  status.smartScanErrorEnabled = computer.computerSettings.antiMalwareSettingSmartScanState;
  return status;
}

/*
 * Finds the Intrusion Prevention rules that protect against a CVE.
 * @param {object} api The api module.
 * @param {Number} cveID The ID of the CVE (for example CVE-2016-7214) or
 *    part of the ID (for example CVE-2016 or 7214).
 * @param {String} apiVersion The api version to use.
 * @return {Promise} A promise that contains the rule IDs.
 */
exports.findRulesForCVE = function(api, cveID, apiVersion) {
  const ruleIDs = [];

  // Search for Intrusion Prevention rules
  return new Promise((resolve, reject) => {
    const getIpRules = () => {
      // Search criteria
      const cveCriteria = new api.SearchCriteria();
      cveCriteria.fieldName = "CVE";
      cveCriteria.stringValue = "%" + cveID + "%";
      cveCriteria.stringTest = api.SearchCriteria.StringTestEnum.equal;

      // Add criteria to a search filter
      const searchFilter = new api.SearchFilter();
      searchFilter.searchCriteria = [cveCriteria];

      // Add the search filter to a search options object
      const searchOptions = {
        searchFilter: searchFilter,
        overrides: false
      };

      // Perform the search
      const ipRulesApi = new api.IntrusionPreventionRulesApi();
      return ipRulesApi.searchIntrusionPreventionRules(apiVersion, searchOptions);
    };

    getIpRules()
      .then(ipRules => {
        // Iterate the rules and get the IDs
        for (let i = 0; i < ipRules.intrusionPreventionRules.length; i++) {
          ruleIDs.push(ipRules.intrusionPreventionRules[i].ID);
        }
        resolve(ruleIDs);
      })
      .catch(error => {
        reject(error);
      });
  });
};

/*
 * Finds computers that are not assigned a specific Intrusion Prevention rule.
 * @param {object} api The api module.
 * @param {Number} ruleID The ID of the rule.
 * @param {String} apiVersion The api version to use.
 * @return {Promise} A promise that contains an array of IDs of unprotected computers.
 */
exports.checkComputersForIPRule = function(api, ruleID, apiVersion) {
  return new Promise((resolve, reject) => {
    // Retrieves computers from Deep Security Manager
    const getComputers = () => {
      const computersApi = new api.ComputersApi();
      return computersApi.listComputers(apiVersion, { overrides: false });
    };

    // Finds computers that are not assigned the rule
    const checkForRule = function(computers) {
      let unprotected = [];
      for (let i = 0; i < computers.computers.length; i++) {
        if (computers.computers[i].intrusionPrevention !== undefined) {
          const IDs = computers.computers[i].intrusionPrevention.ruleIDs;
          let found = false;
          if (IDs !== undefined) {
            for (let j = 0; j < IDs.length; j++) {
              if (IDs[j] === ruleID) {
                found = true;
                break;
              }
            }
          }
          if (!found) {
            unprotected.push(computers.computers[i]);
          }
        }
      }
      return unprotected;
    };

    getComputers()
      .then(computerlist => {
        resolve(checkForRule(computerlist));
      })
      .catch(error => {
        reject(error);
      });
  });
};

/*
 * Assigns an Intrusion Prevention rule to the policy that is assigned to a computer.
 * @param {object} api The api module.
 * @param {Computer} computer The Computer that is assigned a policy.
 * @param {number} ruleID The ID of the Intrusion Prevention rule to add.
 * @param {String} apiVersion The api version to use.
 * @return {Promise} A promise that contains the modified policy.
 */

exports.applyRuleToPolicy = function(api, computer, ruleID, apiVersion) {
  return new Promise((resolve, reject) => {
    // Retrieves the policy that we are modifying
    const getPolicy = policyID => {
      const policiesApi = new api.PoliciesApi();
      return policiesApi.describePolicy(policyID, apiVersion, {
        overrides: false
      });
    };

    // Creates a policy that includes the Intrusion Prevention rules
    // Rule is added to the currently assigned rules so they are not overwritten
    const addRulesToPolicyObject = currentRules => {
      if (currentRules !== undefined) {
        currentRules.push(ruleID);
      } else {
        currentRules = [ruleID];
      }
      const intrusionPreventionPolicyExtension = new api.IntrusionPreventionPolicyExtension();
      intrusionPreventionPolicyExtension.ruleIDs = currentRules;

      const policy = new api.Policy();
      policy.intrusionPrevention = intrusionPreventionPolicyExtension;
      policy.autoRequiresUpdate = api.Policy.AutoRequiresUpdateEnum.on;
      return policy;
    };

    // Updates the policy on Deep Security Manager
    const sendPolicyToManager = (policyID, policy) => {
      const policiesApi = new api.PoliciesApi();
      return policiesApi.modifyPolicy(policyID, policy, apiVersion, {
        overrides: false
      });
    };

    // Check that the computer is assigned a policy
    if (computer.policyID === undefined) {
      reject("Computer has no policy");
      return;
    }
    getPolicy(computer.policyID)
      .then(policy => {
        const newPolicy = addRulesToPolicyObject(policy.intrusionPrevention.ruleIDs);
        resolve(sendPolicyToManager(computer.policyID, newPolicy));
      })
      .catch(error => {
        reject(error);
      });
  });
};

/*
 * Obtains the list of recommended Intrusion Prevention rules to apply to a computer,
 * according to the results of the last recommendation scan.* @param {object} api The api module.
 * @param {object} api The api module.
 * @param {Computer} computer The Computer that was scanned.
 * @param {String} apiVersion The api version to use.
 * @return {Promise} A promise that contains the recommended rules.
 */
exports.getRecommendedIPRules = function(api, computerID, apiVersion) {
  return new Promise((resolve, reject) => {
    // Obtains the results of the recommendation scan
    const getRecommendations = () => {
      const ipRecosApi = new api.ComputerIntrusionPreventionRuleAssignmentsRecommendationsApi();
      return ipRecosApi.listIntrusionPreventionRuleIDsOnComputer(computerID, apiVersion, { overrides: false });
    };

    getRecommendations()
      .then(ipAssignments => {
        // Resolve the recommended rules
        resolve(ipAssignments.assignedRuleIDs);
      })
      .catch(error => {
        reject(error);
      });
  });
};

File: python/src/computer_status_examples.py
# Copyright 2019 Trend Micro.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

def check_anti_malware(api, configuration, api_version, api_exception, computer_id):
    """ Obtains certain anti-malware properties for a computer.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :param computer_id: The ID of the computer.
    :return: An AntiMalwareConfigurationsApi object that contains the anti-malware properties of the computer.
    """

    # Get the computer object from Deep Security Manager
    computers_api = api.ComputersApi(api.ApiClient(configuration))
    computer = computers_api.describe_computer(computer_id, api_version)

    # Get the anti-malware scan configuration id for the computer
    real_time_scan_configuration_id = computer.anti_malware.real_time_scan_configuration_id

    try:
        # Get the anti-malware properties for the computer
        am_configs_api = api.AntiMalwareConfigurationsApi(api.ApiClient(configuration))
        return am_configs_api.describe_anti_malware(real_time_scan_configuration_id, api_version)

    except api_exception as e:
        return "Exception: " + str(e)



def find_rules_for_cve(api, configuration, api_version, api_exception, cve_id):
    """ Finds the intrusion prevention rules for a CVE.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :param cve_id: The ID of the CVE.
    :return: The intrusion prevention rule ID, or None if no rule is found.
    """

    rule_id_s = []

    # Set search criteria
    search_criteria = api.SearchCriteria()
    search_criteria.field_name = "CVE"
    search_criteria.string_value = "%" + cve_id + "%"
    search_criteria.string_test = "equal"

    # Create a search filter
    search_filter = api.SearchFilter()
    search_filter.search_criteria = [search_criteria]

    try:
        # Search for all intrusion prevention rules for the CVE
        ip_rules_api = api.IntrusionPreventionRulesApi(api.ApiClient(configuration))
        ip_rules_search_results = ip_rules_api.search_intrusion_prevention_rules(api_version, search_filter=search_filter)

        # Get the intrusion prevention rule IDs for the CVE from the results
        for rule in ip_rules_search_results.intrusion_prevention_rules:
            rule_id_s.append(rule.id)

        return rule_id_s

    except api_exception as e:
        return "Exception: " + str(e)


def check_computers_for_ip_rule(api, configuration, api_version, api_exception, rule_id):
    """ Finds computers that do not have a specific intrusion prevention rule applied.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :param rule_id: The ID of the rule.
    :return: A list of computers that do not have the rule applied.
    """

    unprotected_computers = []

    try:
        # Create a list of computers
        computers_api = api.ComputersApi(api.ApiClient(configuration))
        computers_list = computers_api.list_computers(api_version, overrides=False)

        # Search the list of computers for those that do not have the IP rule
        for computer in computers_list.computers:
            computer_ip_list = computer.intrusion_prevention
            if computer_ip_list.rule_ids:
                if rule_id in computer_ip_list.rule_ids:
                    unprotected_computers.append(computer)
        return unprotected_computers

    except api_exception as e:
        return "Exception: " + str(e)

def apply_rule_to_policies(api, configuration, api_version, api_exception, computers, rule_id):
    """ Adds an Intrusion Prevention rule to the policies of a list of computers.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :param computer: The Computer that is assigned the policy.
    :param rule_id: The ID of the Intrusion Prevention rule to add.
    :return: A list of PoliciesApi objects that were updated with the rule.
    """


    # Store IDs of policies to modify
    policy_ids = []
    for computer in computers:
        if computer.policy_id:
            policy_ids.append(computer.policy_id)

    # Store modified policies
    modified_policies = []

    for policy_id in policy_ids:
        try:
            # Get the current list of rules from the policy
            policies_api = api.PoliciesApi(api.ApiClient(configuration))
            current_rules = policies_api.describe_policy(policy_id, api_version, overrides=False)

            # Add the rule_id if it doesn't already exist in current_rules
            if current_rules.intrusion_prevention.rule_ids == None:
                current_rules.intrusion_prevention.rule_ids = rule_id

            elif rule_id not in current_rules.intrusion_prevention.rule_ids:
                current_rules.intrusion_prevention.rule_ids.append(rule_id)


            # Add the new and existing intrusion prevention rules to a policy
            intrusion_prevention_policy_extension = api.IntrusionPreventionPolicyExtension()
            intrusion_prevention_policy_extension.rule_ids = current_rules.intrusion_prevention.rule_ids
            policy = api.Policy()
            policy.intrusion_prevention = intrusion_prevention_policy_extension

            # Configure sending policy updates when the policy changes
            policy.auto_requires_update = "on"

            # Modify the policy on Deep Security Manager
            modified_policies.append(policies_api.modify_policy(policy_id, policy, api_version))

        except api_exception as e:
            return "Exception: " + str(e)

    return modified_policies


def get_intrusion_prevention_recommendations(api, configuration, api_version, api_exception, computer_id):
    """Obtains the list of recommended intrusion prevention rules to apply to a computer, according to the results of the last recommendation scan.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :param computer_id: The ID of the computer that was scanned.
    :return: A list of recommended Intrusion Prevention rules to apply to a computer,
    according to the results of the last recommendation scan or None if no scan was performed.
    """

    ip_recommendations_api = api.ComputerIntrusionPreventionRuleAssignmentsRecommendationsApi(api.ApiClient(configuration))
    ip_assignments = None

    try:
        ip_assignments = ip_recommendations_api.list_intrusion_prevention_rule_ids_on_computer(computer_id, api_version, overrides=False)
        return ip_assignments.recommended_to_assign_rule_ids

    except api_exception as e:
        return "Exception: " + str(e)



def get_computer_statuses(api, configuration, api_version, api_exception):
    """Obtains agent and appliance status for all computers and provides the results as comma-separated values.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :return: A string that can be saved as a CSV file.
    """

    # Add column titles to comma-separated values string
    csv = "Host Name,Agent or Appliance,Status,Status Messages,Tasks\r\n"

    try:
        computers_api = api.ComputersApi(api.ApiClient(configuration))
        computers = computers_api.list_computers(api_version, overrides=False)

        for computer in computers.computers:
            computer_info = []

            # Report on computers with no agent or appliance
            if computer.agent_finger_print == None and computer.appliance_finger_print == None:
                # Hostname and protection type
                computer_info.append(computer.host_name)
                computer_info.append("None")

                # Agent/appliance status and status messages
                computer_info.append("No agent/appliance")
                status_messages = ""
                if computer.computer_status != None and computer.computer_status.agent_status != None:
                    status_messages = str(computer.computer_status.agent_status_messages)
                computer_info.append(status_messages)

                # Add the computer info to the CSV string
                csv_line = ""
                for num, item in enumerate(computer_info):
                    csv_line += item
                    if num != (len(computer_info) - 1):
                        csv_line += ","
                    else:
                        csv_line += "\r\n"
                csv += csv_line

            else:
                # Report on problem agents and appliances
                agent_status = computer.computer_status.agent_status
                appliance_status = computer.computer_status.appliance_status

                # Agent is installed but is not active
                if computer.agent_finger_print != None and agent_status != "active":
                    # Hostname and protection type
                    computer_info.append(computer.host_name)
                    computer_info.append("Agent")

                    # Agent status, status messages, and tasks
                    if computer.computer_status.agent_status != None:
                        computer_info.append(computer.computer_status.agent_status)
                    else:
                        computer_info.append("")

                    if computer.computer_status.agent_status_messages != None:
                        computer_info.append(str(computer.computer_status.agent_status_messages))
                    else:
                        computer_info.append("")

                    if computer.tasks != None:
                        computer_info.append(str(computer.tasks.agent_tasks))
                    else:
                        computer_info.append("")

                    # Add the computer info to the CSV string
                    csv_line = ""
                    for num, item in enumerate(computer_info):
                        csv_line += item
                        if num != (len(computer_info) - 1):
                            csv_line += ","
                        else:
                            csv_line += "\r\n"
                    csv += csv_line

                # Appliance is installed but is not active
                if computer.appliance_finger_print != None and appliance_status != "active":
                    # Hostname and protection type
                    computer_info.append(computer.host_name)
                    computer_info.append("Appliance")

                    # Appliance status, status messages, and tasks
                    if computer.computer_status.appliance_status != None:
                        computer_info.append(computer.computer_status.appliance_status)
                    else:
                        computer_info.append("")

                    if computer.computer_status.appliance_status_messages != None:
                        computer_info.append(str(computer.computer_status.appliance_status_messages))
                    else:
                        computer_info.append("")

                    if computer.tasks != None:
                        computer_info.append(str(computer.tasks.appliance_tasks))
                    else:
                        computer_info.append("")

                    # Add the computer info to the CSV string
                    csv_line = ""
                    for num, item in enumerate(computer_info):
                        csv_line += item
                        if num != (len(computer_info) - 1):
                            csv_line += ","
                        else:
                            csv_line += "\r\n"
                    csv += csv_line

        return csv

    except api_exception as e:
        return "Exception: " + str(e)



def get_anti_malware_status_for_computers(api, configuration, api_version, api_exception):
    """Obtains agent and appliance status for the Anti-Malware module of all computers.

    Returns the status information of all computers that have the Anti-Malware module turned off,
    or where the status of the module is not active as comma-separated values.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :return: A string that can be saved as a CSV file.
    """

    # Add column titles to comma-separated values string
    csv = "Host Name,Module State,Agent or Appliance,Status,Status Message\r\n"

    try:
        computers_api = api.ComputersApi(api.ApiClient(configuration))
        computers = computers_api.list_computers(api_version, overrides=False)

        # Get the list of computers and iterate over it
        for computer in computers.computers:
            # Module information to add to the CSV string
            module_info = []

            # Check that the computer has a an agent or appliance status
            if computer.anti_malware.module_status:
                agent_status = computer.anti_malware.module_status.agent_status
                appliance_status = computer.anti_malware.module_status.appliance_status
            else:
                agent_status = None
                appliance_status = None

            # Agents that are not active for the module
            if agent_status and agent_status != "active":
                # Host name
                module_info.append(computer.host_name)

                # Module state
                module_info.append(computer.anti_malware.state)

                # Agent status and status message
                module_info.append("Agent")
                module_info.append(agent_status)
                module_info.append(computer.anti_malware.module_status.agent_status_message)

                # Add the module info to the CSV string
                csv_line = ""
                for num, item in enumerate(module_info):
                    csv_line += item
                    if num != (len(module_info) - 1):
                        csv_line += ","
                    else:
                        csv_line += "\r\n"
                csv += csv_line

            # Appliances that are not active for the module
            if appliance_status and appliance_status != "active":
                # Host name
                module_info.append(computer.host_name)
        
                # Module state
                module_info.append(computer.anti_malware.state)
        
                # Appliance status and status message
                module_info.append("Appliance")
                module_info.append(appliance_status)
                module_info.append(computer.anti_malware.module_status.appliance_status_message)
        
                # Add the module info to the CSV string
                csv_line = ""
                for num, item in enumerate(module_info):
                    csv_line += item
                    if num != (len(module_info) - 1):
                        csv_line += ","
                    else:
                        csv_line += "\r\n"
                csv += csv_line

        return csv

    except api_exception as e:
        return "Exception: " + str(e)
